### PR TITLE
Adds additional escaping throughout plugin (VIP)

### PR DESCRIPTION
**Pull request by tomjn**
There's a number of issues flagged during VIP review, I'm currently going through and fixing them by adding in the missing escaping. WIP PR, questions welcome

### CODE REVIEW

File: includes/class-wp-job-manager-ajax.php [hunk 1]
```diff
@@ -139,7 +139,7 @@ public function get_listings() {
 			'orderby'            => $orderby,
 			'order'              => sanitize_text_field( $_REQUEST['order'] ),
 			'offset'             => ( absint( $_REQUEST['page'] ) - 1 ) * absint( $_REQUEST['per_page'] ),
-			'posts_per_page'     => absint( $_REQUEST['per_page'] ),
+			'posts_per_page'     => max( 1, absint( $_REQUEST['per_page'] ) ),
```
Comment (jom) [inline includes/class-wp-job-manager-ajax.php]: As a behavior change, I might cherry-pick this commit to a separate PR, if that works for you? There could be some themes doing things they shouldn't with `-1`.

Comment (tomjn) [reply includes/class-wp-job-manager-ajax.php]: Sounds good

Comment (jom) [reply includes/class-wp-job-manager-ajax.php]: I started doing this and then realized `absint()` prevented `-1` 😉 . Seems redundant.

Comment (tomjn) [reply includes/class-wp-job-manager-ajax.php]: eh but it does prevent passing `0`, I wanted to add a `min` too so there was an upper limit but i figured that would be problematic

Comment (jom) [reply includes/class-wp-job-manager-ajax.php]: In testing, `0` resulted in the default `per_page`, but oh well.

File: templates/account-signin.php [hunk 2]
```diff
@@ -36,20 +36,20 @@
...
+				<?php printf( esc_html__( 'If you don\'t have an account you can %screate one below by entering your email address/username.', 'wp-job-manager' ), $account_required ? '' : esc_html__( 'optionally', 'wp-job-manager' ) . ' ' ); ?>
 				<?php if ( $use_standard_password_email ) : ?>
-					<?php printf( __( 'Your account details will be confirmed via email.', 'wp-job-manager' ) ); ?>
+					<?php printf( esc_html__( 'Your account details will be confirmed via email.', 'wp-job-manager' ) ); ?>
 				<?php endif; ?>
 
 			<?php elseif ( $account_required ) : ?>
 
-				<?php echo apply_filters( 'submit_job_form_login_required_message',  __('You must sign in to create a new listing.', 'wp-job-manager' ) ); ?>
+				<?php echo wp_kses_post( apply_filters( 'submit_job_form_login_required_message',  __('You must sign in to create a new listing.', 'wp-job-manager' ) ) ); ?>
```
Comment (jom) [inline templates/account-signin.php]: This was an original lint issue, but while editing line, there is a spacing issue in the argument list for `__()`.

**Review by jom — DISMISSED** [review]
Thanks for this PR, @tomjn! Huge improvement. I've added some (mostly linting) notes.

Did some general testing and seems to work well. I've marked it for 1.31.1. I can update the template versions before merging.

**Comment by jom** [comment]
cc: @tripflex I don't _think_ the escaping happening in the `fields/*.php` templates will affect your plugin, but just wanted to flag you.

File: templates/account-signin.php (hunk 2 above)
Comment (tomjn) [reply templates/account-signin.php]: fixed

File: wp-job-manager-template.php [hunk 3]
```diff
@@ -1066,11 +1072,10 @@ function the_company_twitter( $before = '', $after = '', $echo = true, $post = n
 	if ( strlen( $company_twitter ) == 0 )
 		return;
 
-	$company_twitter = esc_attr( strip_tags( $company_twitter ) );
-	$company_twitter = $before . '<a href="http://twitter.com/' . $company_twitter . '" class="company_twitter" target="_blank">' . $company_twitter . '</a>' . $after;
+	$company_twitter = $before . '<a href="'.esc_url( 'https://twitter.com/' . $company_twitter ) . '" class="company_twitter" target="_blank">' . esc_html( wp_strip_all_tags( $company_twitter ) ) . '</a>' . $after;
```
[the remark this answers is not in the document]
Comment (tomjn) [reply wp-job-manager-template.php]: fixed

File: tests/php/bootstrap.php [hunk 4]
```diff
@@ -144,7 +144,7 @@ public function convert_to_exception( $errno, $errstr, $errfile, $errline ) {
 
 		// PHP 5.2 doesn't show the error from Exceptions.
 		if ( version_compare( phpversion(), '5.3.0', '<' ) ) {
-			echo "Error ($errno) - $description\n";
+			echo "Error (".esc_html( $errno ) ."( - ".esc_html( $description )."\n";
```
[the remark this answers is not in the document]
Comment (tomjn) [reply tests/php/bootstrap.php]: fixed

File: templates/job-submit.php [hunk 5]
```diff
@@ -40,7 +40,7 @@
 
 		<?php foreach ( $job_fields as $key => $field ) : ?>
 			<fieldset class="fieldset-<?php echo esc_attr( $key ); ?>">
-				<label for="<?php echo esc_attr( $key ); ?>"><?php echo $field['label'] . apply_filters( 'submit_job_form_required_label', $field['required'] ? '' : ' <small>' . __( '(optional)', 'wp-job-manager' ) . '</small>', $field ); ?></label>
+				<label for="<?php echo esc_attr( $key ); ?>"><?php echo esc_html( $field['label']) . wp_kses_post( apply_filters( 'submit_job_form_required_label', $field['required'] ? '' : ' <small>' . __( '(optional)', 'wp-job-manager' ) . '</small>', $field ) ); ?></label>
```
[the remark this answers is not in the document]
Comment (tomjn) [reply templates/job-submit.php]: fixed

File: templates/job-dashboard.php [hunk 6]
```diff
@@ -28,7 +28,7 @@
 		<tbody>
 			<?php if ( ! $jobs ) : ?>
 				<tr>
-					<td colspan="<?php echo count( $job_dashboard_columns ); ?>"><?php _e( 'You do not have any active listings.', 'wp-job-manager' ); ?></td>
+					<td colspan="<?php echo intval( count( $job_dashboard_columns ) ); ?>"><?php esc_html_e( 'You do not have any active listings.', 'wp-job-manager' ); ?></td>
```
[the remark this answers is not in the document]
Comment (tomjn) [reply templates/job-dashboard.php]: Just being more explicit

File: templates/emails/plain/admin-expiring-job.php [hunk 7]
```diff
@@ -26,12 +26,12 @@
 $expiring_today = $args['expiring_today'];
 
 if ( $expiring_today ) {
-	printf( __( 'The following job listing is expiring today from %s (%s).', 'wp-job-manager' ), get_bloginfo( 'name' ), home_url() );
+	printf( esc_html__( 'The following job listing is expiring today from %s (%s).', 'wp-job-manager' ), esc_html(get_bloginfo( 'name' )), esc_url(home_url()) );
```
[the remark this answers is not in the document]
Comment (tomjn) [reply templates/emails/plain/admin-expiring-job.php]: I tried opening this file and couldn't see the issue, perhaps I fixed it in a previous commit

File: templates/emails/plain/admin-expiring-job.php [hunk 8]
```diff
@@ -26,12 +26,12 @@
...
 } else {
-	printf( __( 'The following job listing is expiring soon from %s (%s).', 'wp-job-manager' ), get_bloginfo( 'name' ), home_url() );
+	printf( esc_html__( 'The following job listing is expiring soon from %s (%s).', 'wp-job-manager' ), esc_html(get_bloginfo( 'name' )), esc_url(home_url()) );
```
[the remark this answers is not in the document]
Comment (tomjn) [reply templates/emails/plain/admin-expiring-job.php]: fixed?

File: templates/emails/employer-expiring-job.php [hunk 9]
```diff
@@ -27,11 +27,14 @@
...
+	echo wp_kses_post(
+		sprintf( __( 'The following job listing is expiring today from <a href="%s">%s</a>.', 'wp-job-manager' ), home_url(), get_bloginfo( 'name' ) ) );
 } else {
-	printf( __( 'The following job listing is expiring soon from <a href="%s">%s</a>.', 'wp-job-manager' ), home_url(), get_bloginfo( 'name' ) );
+	echo wp_kses_post(
+		sprintf( __( 'The following job listing is expiring soon from <a href="%s">%s</a>.', 'wp-job-manager' ), home_url(), get_bloginfo( 'name' ) ) );
 }
-printf( ' ' . __( 'Visit the <a href="%s">job listing dashboard</a> to manage the listing.', 'wp-job-manager' ), esc_url( job_manager_get_permalink( 'job_dashboard' ) ) );
+echo wp_kses_post(
+	sprintf( ' ' . __( 'Visit the <a href="%s">job listing dashboard</a> to manage the listing.', 'wp-job-manager' ), esc_url( job_manager_get_permalink( 'job_dashboard' ) ) ) );
```
[the remark this answers is not in the document]
Comment (tomjn) [reply templates/emails/employer-expiring-job.php]: fixed

File: includes/admin/class-wp-job-manager-cpt-legacy.php [hunk 10]
```diff
@@ -48,18 +48,23 @@ public function __construct() {
 	public function add_bulk_actions_legacy() {
 		global $post_type, $wp_post_types;
 
+		$bulk_actions = [];
```
[the remark this answers is not in the document]
Comment (tomjn) [reply includes/admin/class-wp-job-manager-cpt-legacy.php]: fixed

File: includes/admin/class-wp-job-manager-cpt-legacy.php [hunk 11]
```diff
@@ -48,18 +48,23 @@ public function __construct() {
...
+		foreach( $this->get_bulk_actions() as $key => $bulk_action ) {
+			$bulk_actions[] = [
```
[the remark this answers is not in the document]
Comment (tomjn) [reply includes/admin/class-wp-job-manager-cpt-legacy.php]: fixed

File: includes/admin/class-wp-job-manager-cpt.php [hunk 12]
```diff
@@ -753,17 +755,17 @@ public function extend_submitdiv_post_status() {
...
 		<script type="text/javascript">
 			jQuery( document ).ready( function($) {
 				<?php if ( ! empty( $display ) ) : ?>
-					jQuery( '#post-status-display' ).html( '<?php echo $display; ?>' );
+					jQuery( '#post-status-display' ).html( <?php echo wp_json_encode( $display ); ?> );
 				<?php endif; ?>
 
 				var select = jQuery( '#post-status-select' ).find( 'select' );
-				jQuery( select ).html( "<?php echo $options; ?>" );
+				jQuery( select ).html( <?php echo wp_json_encode($options); ?> );
```
[the remark this answers is not in the document]
Comment (tomjn) [reply includes/admin/class-wp-job-manager-cpt.php]: fixed

File: includes/admin/class-wp-job-manager-writepanels.php [hunk 13]
```diff
@@ -192,33 +192,33 @@ public function job_listing_metabox( $post ) {
...
 			<!-- Display taxonomy terms -->
-			<div id="<?php echo $taxonomy; ?>-all" class="tabs-panel">
-				<ul id="<?php echo $taxonomy; ?>checklist" class="list:<?php echo $taxonomy?> categorychecklist form-no-clear">
+			<div id="<?php echo esc_attr( $taxonomy ); ?>-all" class="tabs-panel">
+				<ul id="<?php echo esc_attr( $taxonomy ); ?>checklist" class="list:<?php echo esc_attr( $taxonomy ); ?> categorychecklist form-no-clear">
 					<?php   foreach($terms as $term){
 						$id = $taxonomy.'-'.$term->term_id;
-						echo "<li id='$id'><label class='selectit'>";
-						echo "<input type='radio' id='in-$id' name='{$name}'".checked($current,$term->term_id,false)."value='$term->term_id' />$term->name<br />";
+						echo '<li id="'.esc_attr( $id )."\"><label class='selectit'>";
```
[the remark this answers is not in the document]
Comment (tomjn) [reply includes/admin/class-wp-job-manager-writepanels.php]: fixed, and I fixed a tonne of other bits that were there from prior

File: includes/admin/class-wp-job-manager-writepanels.php [hunk 14]
```diff
@@ -192,33 +192,33 @@ public function job_listing_metabox( $post ) {
...
-			<div id="<?php echo $taxonomy; ?>-pop" class="tabs-panel" style="display: none;">
-				<ul id="<?php echo $taxonomy; ?>checklist-pop" class="categorychecklist form-no-clear" >
+			<div id="<?php echo esc_attr( $taxonomy ); ?>-pop" class="tabs-panel" style="display: none;">
+				<ul id="<?php echo esc_attr( $taxonomy ); ?>checklist-pop" class="categorychecklist form-no-clear" >
 					<?php   foreach($popular as $term){
 						$id = 'popular-'.$taxonomy.'-'.$term->term_id;
-						echo "<li id='$id'><label class='selectit'>";
-						echo "<input type='radio' id='in-$id'".checked($current,$term->term_id,false)."value='$term->term_id' />$term->name<br />";
+						echo '<li id="'.esc_attr( $id )."\"><label class='selectit'>";
+						echo '<input type="radio" id="in-'.esc_attr( $id ).'" '.checked($current,$term->term_id,false).' value="'.esc_attr( $term->term_id ).'" />'.esc_attr( $term->name ).'<br />';
```
[the remark this answers is not in the document]
Comment (tomjn) [reply includes/admin/class-wp-job-manager-writepanels.php]: fixed

File: templates/job-dashboard.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (jom) [reply templates/job-dashboard.php]: Alright, but `count()` can't return anything but an `int`.

File: assets/js/ajax-filters.js [hunk 15]
```diff
@@ -122,7 +123,8 @@ jQuery( document ).ready( function ( $ ) {
 				if ( result ) {
 					try {
 						if ( result.showing ) {
-							$( showing ).show().html( '<span>' + result.showing + '</span>' + result.showing_links );
+							var showing_el = jQuery('<span>').text(result.showing)
+							$( showing ).show().html( '').text(result.showing_links).prepend(showing_el);
```
Comment (jom) [inline assets/js/ajax-filters.js]: `result.showing_links` is contains HTML and gets encoded using `$.text()`. 

screen shot 2018-06-19 at 4 55 28 pm

**Comment by jom** [comment]
I've force pushed a modified version of that commit with #1503's content removed. I want to test that change with WPML before merging and don't want it to hold up this commit. You'll need to do a hard reset to the remote branch.... sorry, since I'm planning on squashing this PR anyway I could have just done a partial revert commit. 

Can we revert the change to line 125 in `ajax-filters.js`? Once we do that, I'll update the template versions for 1.31.1.

**Comment by tomjn** [comment]
@jom my bandwidth at the moment is quite tight, and we've a client waiting on these changes for quite a while now, anything you can do to expedite things would be super helpful

**Comment by jom** [comment]
@tomjn I'll revert the change mentioned in `ajax-filters.js` so the raw HTML doesn't show up (unless you have another preferred method) and then we should be good for this PR.

**Pull request merged**